File: typescript/src/trace.ts
import { genId, nowIso, utcNow } from "./utils.js";

export enum SpanType {
  LLM_CALL = "llm_call",
  TOOL_CALL = "tool_call",
  TOOL_RESULT = "tool_result",
  REASONING = "reasoning",
  ACTION = "action",
  ERROR = "error",
  FORK_POINT = "fork_point",
}

export enum TraceStatus {
  RUNNING = "running",
  COMPLETED = "completed",
  FAILED = "failed",
}

export interface SpanData {
  id: string;
  trace_id: string;
  parent_id: string | null;
  span_type: string;
  name: string;
  model?: string;
  input?: unknown;
  output?: unknown;
  input_tokens?: number;
  output_tokens?: number;
  cost?: number;
  duration_ms?: number;
  metadata?: Record<string, unknown>;
  started_at: string;
  ended_at?: string;
  error?: string;
}

export interface TraceData {
  id: string;
  name?: string;
  input?: unknown;
  output?: unknown;
  status: string;
  total_tokens: number;
  total_cost: number;
  total_duration_ms: number;
  metadata?: Record<string, unknown>;
  started_at: string;
  ended_at?: string;
  spans?: SpanData[];
  project_id?: string;
}

export class SpanBuilder {
  private data: Partial<SpanData> & { id: string; span_type: string; name: string };
  private _startTime?: Date;

  constructor(name: string, spanType: SpanType) {
    this.data = { id: genId(), span_type: spanType, name, trace_id: "", parent_id: null, started_at: nowIso() };
  }

  setModel(model: string) { this.data.model = model; return this; }
  setInput(input: unknown) { this.data.input = input; return this; }
  setOutput(output: unknown) { this.data.output = output; return this; }
  setParentId(id: string) { this.data.parent_id = id; return this; }
  setTraceId(id: string) { this.data.trace_id = id; return this; }
  setMetadata(m: Record<string, unknown>) { this.data.metadata = m; return this; }

  start(): this {
    this._startTime = utcNow();
    this.data.started_at = this._startTime.toISOString();
    return this;
  }

  end(output?: unknown, error?: string): SpanData {
    const now = utcNow();
    this.data.ended_at = now.toISOString();
    if (output !== undefined) this.data.output = output;
    if (error) this.data.error = error;
    if (this._startTime) {
      this.data.duration_ms = now.getTime() - this._startTime.getTime();
    }
    return this.data as SpanData;
  }

  get id() { return this.data.id; }
  toData(): SpanData { return this.data as SpanData; }
}

export class TraceBuilder {
  private data: TraceData;
  private _startTime?: Date;

  constructor() {
    this.data = {
      id: genId(),
      status: TraceStatus.RUNNING,
      total_tokens: 0,
      total_cost: 0,
      total_duration_ms: 0,
      started_at: nowIso(),
      spans: [],
    };
  }

  start(name?: string, input?: unknown): this {
    this._startTime = utcNow();
    this.data.started_at = this._startTime.toISOString();
    if (name) this.data.name = name;
    if (input !== undefined) this.data.input = input;
    return this;
  }

  end(output?: unknown, status: TraceStatus = TraceStatus.COMPLETED): TraceData {
    const now = utcNow();
    this.data.ended_at = now.toISOString();
    this.data.status = status;
    if (output !== undefined) this.data.output = output;
    if (this._startTime) {
      this.data.total_duration_ms = now.getTime() - this._startTime.getTime();
    }
    return this.data;
  }

  addSpan(span: SpanData) {
    this.data.spans!.push(span);
    this.data.total_tokens += (span.input_tokens || 0) + (span.output_tokens || 0);
    this.data.total_cost += span.cost || 0;
  }

  get id() { return this.data.id; }
  setProjectId(id: string) { this.data.project_id = id; }
  setMetadata(m: Record<string, unknown>) { this.data.metadata = m; }
  toDict(): TraceData { return this.data; }
}

File: python/src/retrace/trace.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Optional
from enum import Enum

from .utils import gen_id, utcnow


class SpanType(str, Enum):
    LLM_CALL = "llm_call"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"
    REASONING = "reasoning"
    ACTION = "action"
    ERROR = "error"
    FORK_POINT = "fork_point"


class TraceStatus(str, Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass
class Span:
    id: str = field(default_factory=gen_id)
    trace_id: str = ""
    span_type: SpanType = SpanType.LLM_CALL
    name: str = ""
    parent_id: Optional[str] = None
    model: Optional[str] = None
    input: Any = None
    output: Any = None
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    cost: Optional[float] = None
    duration_ms: Optional[int] = None
    metadata: dict = field(default_factory=dict)
    started_at: Optional[datetime] = field(default_factory=utcnow)
    ended_at: Optional[datetime] = None
    error: Optional[str] = None

    def to_dict(self) -> dict:
        d: dict[str, Any] = {
            "id": self.id,
            "trace_id": self.trace_id,
            "parent_id": self.parent_id,
            "span_type": self.span_type.value,
            "name": self.name,
            "started_at": self.started_at.isoformat().replace("+00:00", "Z") if self.started_at else None,
        }
        if self.model:
            d["model"] = self.model
        if self.input is not None:
            d["input"] = self.input
        if self.output is not None:
            d["output"] = self.output
        if self.input_tokens is not None:
            d["input_tokens"] = self.input_tokens
        if self.output_tokens is not None:
            d["output_tokens"] = self.output_tokens
        if self.cost is not None:
            d["cost"] = self.cost
        if self.duration_ms is not None:
            d["duration_ms"] = self.duration_ms
        if self.metadata:
            d["metadata"] = self.metadata
        if self.ended_at:
            d["ended_at"] = self.ended_at.isoformat().replace("+00:00", "Z")
        if self.error:
            d["error"] = self.error
        return d


@dataclass
class Trace:
    id: str = field(default_factory=gen_id)
    name: Optional[str] = None
    input: Any = None
    output: Any = None
    status: TraceStatus = TraceStatus.RUNNING
    total_tokens: int = 0
    total_cost: float = 0.0
    total_duration_ms: int = 0
    metadata: dict = field(default_factory=dict)
    started_at: Optional[datetime] = field(default_factory=utcnow)
    ended_at: Optional[datetime] = None
    spans: list = field(default_factory=list)
    project_id: Optional[str] = None

    def to_dict(self) -> dict:
        d: dict[str, Any] = {
            "id": self.id,
            "status": self.status.value,
            "total_tokens": self.total_tokens,
            "total_cost": self.total_cost,
            "total_duration_ms": self.total_duration_ms,
            "started_at": self.started_at.isoformat().replace("+00:00", "Z") if self.started_at else None,
        }
        if self.name:
            d["name"] = self.name
        if self.input is not None:
            d["input"] = self.input
        if self.output is not None:
            d["output"] = self.output
        if self.metadata:
            d["metadata"] = self.metadata
        if self.ended_at:
            d["ended_at"] = self.ended_at.isoformat().replace("+00:00", "Z")
        if self.project_id:
            d["project_id"] = self.project_id
        if self.spans:
            d["spans"] = [s.to_dict() for s in self.spans]
        return d
